src/arcam/fmj/utils.py

```diff
@@ -56,33 +60,69 @@ def get_uniqueid_from_udn(data) -> Optional[str]:
         return None
 
 
-def get_possibly_invalid_xml(data) -> Any:
-    try:
-        return ElementTree.fromstring(data)
-    except ElementTree.ParseError:
-        _LOGGER.info("Device provide corrupt xml, trying with ampersand replacement")
-        data = re.sub(r'&(?![A-Za-z]+[0-9]*;|#[0-9]+;|#x[0-9a-fA-F]+;)', r'&amp;', data)
-        return ElementTree.fromstring(data)
+async def get_upnp_headers(host: str) -> Optional[CaseInsensitiveDict]:
+    """Get search response headers from a host based on ssdp/upnp."""
+    search_target = "upnp:rootdevice"
 
-def get_udn_from_xml(xml: Any) -> Optional[str]:
-    return xml.findtext("d:device/d:UDN", None, {"d": "urn:schemas-upnp-org:device-1-0"})
+    responses : List[CaseInsensitiveDict] = []
 
-async def get_uniqueid_from_device_description(session: aiohttp.ClientSession, url: str):
-    """Retrieve and extract unique id from url."""
-    try:
-        async with session.get(url) as req:
-            req.raise_for_status()
-            data = await req.text()
-            xml = get_possibly_invalid_xml(data)
-            udn = get_udn_from_xml(xml)
-            return get_uniqueid_from_udn(udn)
-    except (aiohttp.ClientError, asyncio.TimeoutError, ElementTree.ParseError):
-        _log_exception("Unable to get device description from %s", url)
+    async def _handle_response(headers: CaseInsensitiveDict) -> None:
+        responses.append(headers)
+
+    await async_search(
+        search_target=search_target,
+        target=(host, SSDP_PORT),
+        async_callback=_handle_response,
+    )
+
+    if len(responses) == 0:
+        _LOGGER.warning(f"No UPNP response from {host}")
+        return None
+    elif len(responses) > 1:
+        _LOGGER.warning(f"More than one UPNP response from {host}")
```

Comment:
why is this an issue?

Reply:
I'm perhaps being over-rigorous but I'm only expecting one response because I specifically ask for the `upnp:rootdevice`.    I wasn't sure under what circumstances I might get multiple responses and it hasn't actually happened as far as I know.

That said, I could change it to `return responses[0]` if you prefer.   I'd suggest to keep the warning though.